src/rt-app_types.h

```diff
@@ -206,6 +212,10 @@ typedef struct _thread_data_t {
 
 	int forked;
 	int num_instances;
+
+	struct _rtapp_options_t *opts;
+	rtapp_resources_t *local_resources;
+	rtapp_resources_t **global_resources;
```

Comment:
Does gloabal_resources need to be a double pointer? It seems to me an artefact left over from previous code?

Reply:
yes it has to be double pointer
because realloc can modify the pointer, we must be able to save the new pointer.

Reply:
Maybe it's friday afternoon that's clouding my thinking. But we can save the returned pointer with a single pointer too. What am I missing? When forking we use realloc to expand `threads` structure - which is a single pointer. If we fail to expand then we basically fail to run the whole test and bail out.

Reply:
realloc can move the  struct somewhere else so we must update the place where the pointer is saved. In fact, this happened during some tests.
If you look at add_resource_data(), we don't use the local pointer table but *resources_table after the realloc

Reply:
>If you look at add_resource_data(), we don't use the local pointer table but *resources_table after the realloc

Yes because it's a double pointer. If you make it a single pointer you can just access it directly, no?

Because you added rtapp_resources_t there's no need for the double pointer AFAICS and I don't see why local_resources can be a single pointer but not global_resources since the table is now embedded in the new struct.

We expand threads in fork event in rt-app.c using realloc, why the usage of realloc in the 2 cases is not equivalent?

Reply:
no, otherwise we would only update the "local" pointer of gloabl_resources of the task that has added the resources but other tasks will continue to point to previous one.
single pointer is enough for loacl_resources because it's not point by others.

we have opts.resources->the_global_resouces and threadA_data.global_resources-> opts.resources->the_global_resouces and threadB_data.global_resources-> opts.resources->the_global_resouces. If threadA_data and threadB_data were directly pointing to the_global_resouces instead, if @ of the_global_resouces changes, only the thread that did the changes would have the new @ others would still point to previous @

Reply:
Oh I see what I was missing now. This is a per thread pointer to a global resource. So we are saving a pointer to a pointer actually - I was thinking the table stored directly in this struct and missed it's part of the tread data, my bad.

If we already have a pointer to the global_resources in opts then we should just use that IMHO. It's very easy to create a micro to improve readability. But I don't feel strongly about it.

Reply:
This pointer is used quite often when running events and make the code more readable instead of finding our way to the pointer every time.

Reply:
opts pointer is now removed from thread structure because it was only used for resources init like pi_enable